File: bindings/java/src-completable/test/com/apple/cie/foundationdb/test/RYWBenchmark.java
/*
 * RYWBenchmark.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.cie.foundationdb.test;

import com.apple.cie.foundationdb.Database;
import com.apple.cie.foundationdb.Transaction;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class RYWBenchmark extends AbstractTester {
    private int keyCount;

    public static final int DEFAULT_KEY_COUNT = 10_000;
    public static final int DEFAULT_KEY_SIZE = 16;

    private final Map<String, Function<? super Transaction, ? extends Double>> tests;
    private final String keyFormat;

    public RYWBenchmark() {
        this(DEFAULT_KEY_COUNT, DEFAULT_KEY_SIZE);
    }

    public RYWBenchmark(int keyCount, int keySize) {
        super();
        this.keyCount = keyCount;

        keyFormat = "%0" + keySize + "d";

        tests = new HashMap<>();
        tests.put("get_single", tr -> getSingle(tr, 10_000));
        tests.put("get_many_sequential", tr -> getManySequential(tr, 10_000));
        tests.put("get_range_basic", tr -> getRangeBasic(tr, 1_000));
        tests.put("single_clear_get_range", tr -> singleClearGetRange(tr, 1_000));
        tests.put("clear_range_get_range", tr -> clearRangeGetRange(tr, 1_000));
        tests.put("interleaved_sets_gets", tr -> interleavedSetsGets(tr, 10_000));
    }

    @Override
    public void testPerformance(Database db) {
        Transaction tr = db.createTransaction();
        insertData(tr);

        List<String> testsToRun;
        if (args.getTestsToRun().isEmpty()) {
            testsToRun = tests.keySet().stream().sorted().collect(Collectors.toList());
        } else {
            testsToRun = args.getTestsToRun();
        }

        for (String test : testsToRun) {
            if (!tests.containsKey(test)) {
                result.addError(new IllegalArgumentException("Test " + test + " not implemented"));
                continue;
            }

            Function<? super Transaction, ? extends Double> function = tests.get(test);

            try {
                Thread.sleep(5_000);
            } catch (InterruptedException e) {
                result.addError(wrapAndPrintError(e, "Interrupted while sleeping"));
            }

            System.out.println("Running test " + test);

            List<Double> results = new ArrayList<>(NUM_RUNS);

            for (int i = 0; i < NUM_RUNS; i++) {
                try {
                    results.add(function.apply(tr));
                } catch (Exception e) {
                    result.addError(wrapAndPrintError(e, "Performance test failed: " + test));
                    break;
                }
            }

            if (results.size() == NUM_RUNS) {
                Collections.sort(results);
                result.addKpi(String.format("%s (%s)", test, multiVersionDescription()), results.get(results.size() / 2).intValue(), "keys/s");
            }
        }

        tr.cancel();
    }

    public Double getSingle(Transaction tr, int count) {
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
            tr.get(key(5001)).join();
        }
        long end = System.nanoTime();

        return count*1_000_000_000.0/(end - start);
    }

    public Double getManySequential(Transaction tr, int count) {
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
            tr.get(key(i)).join();
        }
        long end = System.nanoTime();

        return count*1_000_000_000.0/(end - start);
    }

    public Double getRangeBasic(Transaction tr, int count) {
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
           tr.getRange(key(0), key(keyCount)).asList().join();
        }
        long end = System.nanoTime();

        return count * 1_000_000_000.0 * keyCount/(end - start);
    }

    public Double singleClearGetRange(Transaction tr, int count) {
        for (int i = 0; i < keyCount; i += 2) {
            tr.clear(("" + i).getBytes(ASCII));
        }
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
            tr.getRange(key(0), key(keyCount)).asList().join();
        }
        long end = System.nanoTime();

        Double kpi = count * 1_000_000_000.0 * keyCount/(end - start);
        insertData(tr);
        return kpi;
    }

    public Double clearRangeGetRange(Transaction tr, int count) {
        for (int i = 0; i < keyCount; i += 4) {
            tr.clear(key(i), key(i+1));
        }
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
            tr.getRange(key(0), key(keyCount)).asList().join();
        }
        long end = System.nanoTime();

        Double kpi = count * 1_000_000_000.0 * keyCount/(end - start);
        insertData(tr);
        return kpi;
    }

    public Double interleavedSetsGets(Transaction tr, int count) {
        long start = System.nanoTime();
        byte[] keyBytes = "foo".getBytes(ASCII);
        tr.set(keyBytes, "1".getBytes(ASCII));
        for (int i = 0; i < count; i++) {
            int old = Integer.parseInt(new String(tr.get(keyBytes).join(), ASCII));
            tr.set(keyBytes, ("" + (old + 1)).getBytes(ASCII));
        }
        long end = System.nanoTime();

        return count * 1_000_000_000.0/(end - start);
    }

    public void insertData(Transaction tr) {
        tr.clear(new byte[0], new byte[]{(byte)0xff}); // Clear user space.
        for (int i = 0; i < keyCount; i++) {
            tr.set(key(i), "foo".getBytes(ASCII));
        }
    }

    public byte[] key(int i) {
        return String.format(keyFormat, i).getBytes(ASCII);
    }

    public static void main(String[] args) {
        System.out.println("Running Java RYW benchmark on Java version " + System.getProperty("java.version"));
        try {
            new RYWBenchmark().run(args);
        } catch (IllegalArgumentException e) {
            System.out.println("Could not run test due to malformed arguments.");
            System.exit(1);
        } catch (Exception e) {
            System.out.println("Fatal error encountered during run: " + e);
            e.printStackTrace();
            System.exit(2);
        }
    }
}

File: bindings/c/test/ryw_benchmark.c
/*
 * ryw_benchmark.c
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "test.h"
#include <foundationdb/fdb_c.h>
#include <foundationdb/fdb_c_options.g.h>

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

pthread_t netThread;

int numKeys = 10000;
int keySize = 16;
uint8_t** keys;

void insertData(FDBTransaction* tr) {
	fdb_transaction_clear_range(tr, (uint8_t*)"", 0, (uint8_t*)"\xff", 1);

	uint8_t* v = (uint8_t*)"foo";
	uint32_t i;
	for (i = 0; i <= numKeys; ++i) {
		fdb_transaction_set(tr, keys[i], keySize, v, 3);
	}
}

int runTest(int (*testFxn)(FDBTransaction*, struct ResultSet*),
            FDBTransaction* tr,
            struct ResultSet* rs,
            const char* kpiName) {
	int numRuns = 25;
	int* results = malloc(sizeof(int) * numRuns);
	int i = 0;
	for (; i < numRuns; ++i) {
		results[i] = testFxn(tr, rs);
		if (results[i] < 0) {
			free(results);
			return -1;
		}
	}

	int result = median(results, numRuns);
	free(results);

	addKpi(rs, kpiName, result, "keys/s");

	return result;
}

int getSingle(FDBTransaction* tr, struct ResultSet* rs) {
	int present;
	uint8_t const* value;
	int length;
	int i;

	double start = getTime();
	for (i = 0; i < numKeys; ++i) {
		FDBFuture* f = fdb_transaction_get(tr, keys[5001], keySize, 0);
		if (getError(fdb_future_block_until_ready(f), "GetSingle (block for get)", rs))
			return -1;
		if (getError(fdb_future_get_value(f, &present, &value, &length), "GetSingle (get result)", rs))
			return -1;
		fdb_future_destroy(f);
	}
	double end = getTime();

	return numKeys / (end - start);
}

int getManySequential(FDBTransaction* tr, struct ResultSet* rs) {
	int present;
	uint8_t const* value;
	int length;
	int i;

	double start = getTime();
	for (i = 0; i < numKeys; ++i) {
		FDBFuture* f = fdb_transaction_get(tr, keys[i], keySize, 0);
		if (getError(fdb_future_block_until_ready(f), "GetManySequential (block for get)", rs))
			return -1;
		if (getError(fdb_future_get_value(f, &present, &value, &length), "GetManySequential (get result)", rs))
			return -1;
		fdb_future_destroy(f);
	}
	double end = getTime();

	return numKeys / (end - start);
}

int getRangeBasic(FDBTransaction* tr, struct ResultSet* rs) {
	int count;
	const FDBKeyValue* kvs;
	int more;
	int i;

	double start = getTime();
	for (i = 0; i < 100; ++i) {
		FDBFuture* f = fdb_transaction_get_range(tr,
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[0], keySize),
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[numKeys], keySize),
		                                         numKeys,
		                                         0,
		                                         0,
		                                         1,
		                                         0,
		                                         0);

		if (getError(fdb_future_block_until_ready(f), "GetRangeBasic (block for get range)", rs))
			return -1;
		if (getError(fdb_future_get_keyvalue_array(f, &kvs, &count, &more), "GetRangeBasic (get range results)", rs))
			return -1;

		if (count != numKeys) {
			fprintf(stderr, "Bad count %d (expected %d)\n", count, numKeys);
			addError(rs, "GetRangeBasic bad count");
			return -1;
		}
	}
	double end = getTime();

	return 100 * numKeys / (end - start);
}

int singleClearGetRange(FDBTransaction* tr, struct ResultSet* rs) {
	int count;
	const FDBKeyValue* kvs;
	int more;
	int i;

	for (i = 0; i < numKeys; i += 2) {
		fdb_transaction_clear(tr, keys[i], keySize);
	}

	double start = getTime();
	for (i = 0; i < 100; ++i) {
		FDBFuture* f = fdb_transaction_get_range(tr,
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[0], keySize),
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[numKeys], keySize),
		                                         numKeys,
		                                         0,
		                                         0,
		                                         1,
		                                         0,
		                                         0);

		if (getError(fdb_future_block_until_ready(f), "SingleClearGetRange (block for get range)", rs))
			return -1;
		if (getError(
		        fdb_future_get_keyvalue_array(f, &kvs, &count, &more), "SingleClearGetRange (get range results)", rs))
			return -1;

		fdb_future_destroy(f);

		if (count != numKeys / 2) {
			fprintf(stderr, "Bad count %d (expected %d)\n", count, numKeys);
			addError(rs, "SingleClearGetRange bad count");
			return -1;
		}
	}
	double end = getTime();

	insertData(tr);
	return 100 * numKeys / 2 / (end - start);
}

int clearRangeGetRange(FDBTransaction* tr, struct ResultSet* rs) {
	int count;
	const FDBKeyValue* kvs;
	int more;
	int i;

	for (i = 0; i < numKeys; i += 4) {
		fdb_transaction_clear_range(tr, keys[i], keySize, keys[i + 1], keySize);
	}

	double start = getTime();
	for (i = 0; i < 100; ++i) {
		FDBFuture* f = fdb_transaction_get_range(tr,
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[0], keySize),
		                                         FDB_KEYSEL_LAST_LESS_OR_EQUAL(keys[numKeys], keySize),
		                                         numKeys,
		                                         0,
		                                         0,
		                                         1,
		                                         0,
		                                         0);

		if (getError(fdb_future_block_until_ready(f), "ClearRangeGetRange (block for get range)", rs))
			return -1;
		if (getError(
		        fdb_future_get_keyvalue_array(f, &kvs, &count, &more), "ClearRangeGetRange (get range results)", rs))
			return -1;

		fdb_future_destroy(f);

		if (count != numKeys * 3 / 4) {
			fprintf(stderr, "Bad count %d (expected %d)\n", count, numKeys * 3 / 4);
			addError(rs, "ClearRangeGetRange bad count");
			return -1;
		}
	}
	double end = getTime();

	insertData(tr);
	return 100 * numKeys * 3 / 4 / (end - start);
}

int interleavedSetsGets(FDBTransaction* tr, struct ResultSet* rs) {
	int present;
	uint8_t const* value;
	int length;
	int i;

	uint8_t* k = (uint8_t*)"foo";
	uint8_t v[10];
	int num = 1;

	double start = getTime();
	sprintf((char*)v, "%d", num);
	fdb_transaction_set(tr, k, 3, v, strlen((char*)v));

	for (i = 0; i < 10000; ++i) {
		FDBFuture* f = fdb_transaction_get(tr, k, 3, 0);
		if (getError(fdb_future_block_until_ready(f), "InterleavedSetsGets (block for get)", rs))
			return -1;
		if (getError(fdb_future_get_value(f, &present, &value, &length), "InterleavedSetsGets (get result)", rs))
			return -1;
		fdb_future_destroy(f);

		sprintf((char*)v, "%d", ++num);
		fdb_transaction_set(tr, k, 3, v, strlen((char*)v));
	}
	double end = getTime();

	return 10000 / (end - start);
}

void runTests(struct ResultSet* rs) {
	FDBDatabase* db = openDatabase(rs, &netThread);

	FDBTransaction* tr;
	checkError(fdb_database_create_transaction(db, &tr), "create transaction", rs);

	FDBFuture* f = fdb_transaction_get_read_version(tr);
	checkError(fdb_future_block_until_ready(f), "block for read version", rs);

	int64_t version;
	checkError(fdb_future_get_int64(f, &version), "get version", rs);
	fdb_future_destroy(f);

	insertData(tr);

	runTest(&getSingle, tr, rs, "C: get single cached value throughput");
	runTest(&getManySequential, tr, rs, "C: get sequential cached values throughput");
	runTest(&getRangeBasic, tr, rs, "C: get range cached values throughput");
	runTest(&singleClearGetRange, tr, rs, "C: get range cached values with clears throughput");
	runTest(&clearRangeGetRange, tr, rs, "C: get range cached values with clear ranges throughput");
	runTest(&interleavedSetsGets, tr, rs, "C: interleaved sets and gets on a single key throughput");

	fdb_transaction_destroy(tr);
	fdb_database_destroy(db);
	fdb_stop_network();
}

int main(int argc, char** argv) {
	srand(time(NULL));
	struct ResultSet* rs = newResultSet();
	checkError(fdb_select_api_version(720), "select API version", rs);
	printf("Running RYW Benchmark test at client version: %s\n", fdb_get_client_version());

	keys = generateKeys(numKeys, keySize);
	runTests(rs);
	writeResultSet(rs);
	freeResultSet(rs);
	freeKeys(keys, numKeys);

	return 0;
}
